File: Sources/Store/ReminderPlanner.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import Foundation
import UserNotifications

/// Wipe-and-reschedule local-notification planning for the daily ritual.
///
/// Deterministic `doigt.routine.<uuid>.<slot>` identifiers plus a full replan on every
/// change make scheduling idempotent — there is no incremental state to corrupt, and
/// re-adding an identifier replaces the request in place, so editing 08:00 → 09:00
/// moves one reminder rather than accumulating two.
///
/// Satisfied slots are suppressed for TODAY only — the TRAINING day, which turns at
/// `DayStamp.rolloverHour`, because that is the day `outstandingToday` was counted for.
/// iOS has no start date for a daily repeating calendar trigger, so we fill its
/// available 64-request budget with dated one-shot requests, earliest first. Normal app
/// activity replenishes this finite horizon; opening the app is required before the
/// queued horizon runs out.
@MainActor
enum ReminderPlanner {
    /// One routine's reminder settings, flattened to Sendable value data so the whole
    /// plan can be computed off the main actor without touching a `@Model` object.
    struct RoutinePlanInput: Hashable, Sendable {
        let id: UUID
        let name: String
        let reminders: [ReminderTime]
        let enabled: Bool
        /// How many sessions today still owes. Zero means the day is already met and
        /// every one of this routine's remaining slots is suppressed.
        ///
        /// A reminder that fires after the day's sessions is the app failing to notice
        /// you succeeded — the thing that gets notifications turned off for good.
        var outstandingToday: Int = 1
    }

    /// A request, fully resolved but not yet handed to the system — which is what makes
    /// the planning testable without a notification center.
    struct PlannedReminder: Hashable, Sendable {
        let identifier: String
        let title: String
        let body: String
        let components: DateComponents
        var suppressToday: Bool = false
    }

    /// The namespace we own. Everything with this prefix is ours to delete on a
    /// replan; anything without it belongs to another feature and is left alone.
    /// `nonisolated` so the pure half of this enum stays callable from anywhere.
    nonisolated static let identifierPrefix = "doigt.routine."

    /// Content-keyed on both halves: the routine's UUID survives an undo-delete, and
    /// `slot` is derived from the TIME, so a moved slot replaces its own request
    /// rather than leaving an orphan firing at the old hour.
    nonisolated static func identifier(routine: UUID, slot: ReminderTime) -> String {
        "\(identifierPrefix)\(routine.uuidString).\(slot.slot)"
    }

    /// The whole plan, as a pure function of the routines — no clock, no notification
    /// center, no authorization. Sorted and deduped, and a routine with reminders
    /// switched off contributes nothing rather than contributing a disabled request.
    nonisolated static func requests(for routines: [RoutinePlanInput]) -> [PlannedReminder] {
        var planned: [PlannedReminder] = []
        // Two routines cannot share an identifier (the UUID is in it), but a caller
        // that passes the same routine twice must not produce a duplicate request.
        var claimed: Set<String> = []

        for routine in routines where routine.enabled {
            // Suppress from the FRONT of the TRAINING day: having trained once, the
            // morning slot is satisfied and the evening one still owed. A 01:00 slot is
            // the last of the evening before, so it sorts after 23:00.
            let sorted = Set(routine.reminders).sorted {
                $0.trainingDayOrder < $1.trainingDayOrder
            }
            let suppressed = max(0, sorted.count - max(0, routine.outstandingToday))
            for (index, slot) in sorted.enumerated() {
                let id = identifier(routine: routine.id, slot: slot)
                guard claimed.insert(id).inserted else { continue }
                planned.append(PlannedReminder(
                    identifier: id,
                    title: routine.name,
                    body: String(localized: "Time for a session."),
                    components: slot.dateComponents,
                    suppressToday: index < suppressed
                ))
            }
        }
        return planned
    }

    /// Resolve the slot plan to future calendar dates. Every enabled slot returns
    /// tomorrow even when today's target is met.
    ///
    /// **Suppression is keyed to the TRAINING day each firing falls in**, never the
    /// calendar day: a session at 00:30 satisfied the evening before, and must not delete
    /// the next morning's reminders. A 01:00 slot after a 23:30 session is still that
    /// evening, so it is silenced.
    nonisolated static func scheduledRequests(for routines: [RoutinePlanInput],
                                              now: Date = .now,
                                              calendar: Calendar = .current,
                                              limit: Int = 64) -> [PlannedReminder] {
        let slots = requests(for: routines)
        guard !slots.isEmpty, limit > 0 else { return [] }
        let budget = min(64, limit)
        let today = calendar.startOfDay(for: now)
        let trainingToday = DayStamp(trainingDayOf: now, calendar: calendar)
        var result: [PlannedReminder] = []
        for offset in 0...budget {
            guard let day = calendar.date(byAdding: .day, value: offset, to: today) else { continue }
            var candidates: [(Date, PlannedReminder)] = []
            for slot in slots {
                guard let hour = slot.components.hour, let minute = slot.components.minute,
                      let fire = calendar.date(bySettingHour: hour, minute: minute, second: 0,
                                               of: day), fire > now,
                      calendar.isDate(fire, inSameDayAs: day) else { continue }
                if slot.suppressToday,
                   DayStamp(trainingDayOf: fire, calendar: calendar) == trainingToday { continue }
                let components = calendar.dateComponents([.year, .month, .day, .hour, .minute, .second],
                                                         from: fire)
                let dated = PlannedReminder(
                    identifier: "\(slot.identifier).d\(DayStamp(date: day, calendar: calendar).raw)",
                    title: slot.title, body: slot.body, components: components)
                candidates.append((fire, dated))
            }
            candidates.sort {
                $0.0 == $1.0 ? $0.1.identifier < $1.1.identifier : $0.0 < $1.0
            }
            result.append(contentsOf: candidates.prefix(budget - result.count).map { $0.1 })
            if result.count == budget { break }
        }
        return result
    }

    /// Replans are fully serialized: each new one cancels its predecessor AND awaits it
    /// before touching the notification center, so a superseded run's in-flight adds can
    /// never land after the successor's wipe.
    private static var inflight: Task<Void, Never>?

    static func replan(_ routines: [RoutinePlanInput]) async {
        let predecessor = inflight
        predecessor?.cancel()
        // DETACHED: a plain `Task {}` inherits this enum's main actor, which in the
        // sibling app cost 70–200 ms on the main thread after every save. The worker
        // touches only Sendable values and the thread-safe UNUserNotificationCenter.
        let task = Task.detached(priority: .utility) {
            _ = await predecessor?.value
            await run(routines)
        }
        inflight = task
        await task.value
    }

    private nonisolated static func run(_ routines: [RoutinePlanInput]) async {
        #if DEBUG
        assert(!Thread.isMainThread, "reminder planning must stay off the main thread")
        let computeStart = ContinuousClock.now
        #endif
        let center = UNUserNotificationCenter.current()

        // Enumerate first so unrelated notifications retain their budget. Equivalent
        // dated requests are replaced in place; only obsolete owned IDs are retired.
        let pending = await center.pendingNotificationRequests()
        guard !Task.isCancelled else { return }
        let ours = Set(pending.map(\.identifier).filter { $0.hasPrefix(identifierPrefix) })

        guard !requests(for: routines).isEmpty else {
            // Every routine's reminders are off (or there are no routines): the correct
            // plan is genuinely empty, and this is the one path that may clear without
            // checking authorization.
            center.removePendingNotificationRequests(withIdentifiers: Array(ours))
            return
        }

        let status = await center.notificationSettings().authorizationStatus
        guard status == .authorized || status == .provisional, !Task.isCancelled else { return }

        // Other features retain their requests and their share of the system budget.
        let available = max(0, 64 - (pending.count - ours.count))
        let planned = scheduledRequests(for: routines, limit: available)
        guard !planned.isEmpty else {
            // Even with no free budget, an obsolete repeater must not keep nagging
            // today after completion. Requests owned by other features stay untouched.
            center.removePendingNotificationRequests(withIdentifiers: Array(ours))
            return
        }
        let wanted = Set(planned.map(\.identifier))
        // Retire only obsolete requests before adding: otherwise the migration from
        // repeaters, or a refreshed full horizon, can temporarily exceed iOS's budget.
        center.removePendingNotificationRequests(withIdentifiers: Array(ours.subtracting(wanted)))
        for item in planned {
            guard !Task.isCancelled else { return }
            let content = UNMutableNotificationContent()
            content.title = item.title
            content.body = item.body
            content.sound = .default
            // `.active`, not `.timeSensitive`: a habit nudge must not pierce Focus.
            content.interruptionLevel = .active

            let trigger = UNCalendarNotificationTrigger(dateMatching: item.components, repeats: false)
            do {
                try await center.add(UNNotificationRequest(identifier: item.identifier,
                                                          content: content, trigger: trigger))
            } catch {
                // Existing requests with this dated identifier remain installed if a
                // replacement fails; the next normal replan retries the missing add.
                continue
            }
        }

        #if DEBUG
        if ProcessInfo.processInfo.arguments.contains("-timings") {
            print("[timings] reminder plan \(ContinuousClock.now - computeStart) off-main (\(planned.count) items)")
        }
        #endif

    }

    /// Contextual, one-shot permission ask: the first Save of a routine WITH reminders
    /// on is the moment notifications become meaningful. Never at cold launch, where
    /// the alert arrives before the user knows what the app does.
    static func requestAuthorizationIfNeeded(settings: SettingsStore) async {
        guard !settings.didAskNotificationPermission else { return }
        // Set BEFORE the await: two quick saves would otherwise stack two OS alerts.
        settings.didAskNotificationPermission = true
        _ = try? await UNUserNotificationCenter.current()
            .requestAuthorization(options: [.alert, .sound, .badge])
    }
}

File: android/app/src/main/kotlin/run/nuri/getagrip/store/ReminderPlanner.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.store

import android.Manifest
import android.app.AlarmManager
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import androidx.core.app.NotificationCompat
import androidx.core.content.ContextCompat
import run.nuri.getagrip.MainActivity
import run.nuri.getagrip.R
import run.nuri.getagrip.engine.DayStamp
import run.nuri.getagrip.engine.L10n
import run.nuri.getagrip.engine.ReminderTime
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.UUID

/// Wipe-and-reschedule local-notification planning for the daily ritual.
///
/// Deterministic `doigt.routine.<uuid>.<slot>` identifiers plus a full replan on every
/// change make scheduling idempotent: no incremental state to corrupt, and re-adding an
/// identifier replaces the alarm in place, so 08:00 → 09:00 moves one reminder rather than
/// adding a second.
///
/// TRANSLATION NOTE (Sources/Store/ReminderPlanner.swift): `requests(for:)`, the pure half
/// the tests exercise, is unchanged in its rules. Underneath:
///
/// - `UNUserNotificationCenter` becomes `AlarmManager` + a `BroadcastReceiver`.
///   **`setWindow` with a 15-minute window, never `setExactAndAllowWhileIdle`**: exact
///   alarms need `SCHEDULE_EXACT_ALARM`, which Play gates to alarm-clock and calendar apps,
///   and a habit nudge insisting on the minute is the overreach `.timeSensitive` was on
///   iOS.
/// - iOS's repeating trigger becomes a ONE-SHOT alarm the receiver reschedules after
///   firing. `setRepeating` is inexact and un-cancellable per occurrence; rescheduling on
///   fire lets a session at 18:00 suppress TODAY's 19:00 slot and restore it tomorrow.
/// - The identifier is hashed into a stable request code (`requestCode`) naming the
///   `PendingIntent`.
object ReminderPlanner {

    /// One routine's reminder settings as value data, so the plan is computed without a
    /// database row.
    data class RoutinePlanInput(
        val id: UUID,
        val name: String,
        val reminders: List<ReminderTime>,
        val enabled: Boolean,
        /// How many sessions today still owes. Zero means the day is met and every
        /// remaining slot is suppressed: a reminder after you have done the thing is how
        /// notifications get turned off for good.
        val outstandingToday: Int = 1,
    )

    /// A request fully resolved but not yet handed to the system, so planning is testable
    /// without `AlarmManager`. `hour`/`minute` stand in for the only two `DateComponents`
    /// fields the iOS trigger set.
    data class PlannedReminder(
        val identifier: String,
        val title: String,
        val body: String,
        val hour: Int,
        val minute: Int,
        val suppressToday: Boolean = false,
    ) {
        /// `AlarmManager` addresses an alarm by `PendingIntent`, identified by request
        /// code, so the identifier collapses to an Int. `String.hashCode` is specified
        /// (stable across processes), and the collision domain is this app's own alarms.
        val requestCode: Int get() = identifier.hashCode()
    }

    /// The namespace we own: everything with this prefix is ours to cancel on a replan.
    /// Kept as `doigt.` (the iOS identity, shared vocabulary between the ports, never
    /// user-visible).
    const val identifierPrefix = "doigt.routine."

    /// Content-keyed on both halves: the routine's UUID survives an undo-delete, and `slot`
    /// derives from the TIME, so moving 08:00 to 09:00 replaces its own request rather than
    /// leaving an orphan firing forever.
    fun identifier(routine: UUID, slot: ReminderTime): String =
        "$identifierPrefix${routine.toString().uppercase()}.${slot.slot}"

    /// The whole plan as a pure function of the routines — no clock, notification manager
    /// or permission. Sorted and deduped; a routine with reminders off contributes nothing.
    fun requests(routines: List<RoutinePlanInput>): List<PlannedReminder> {
        val planned = mutableListOf<PlannedReminder>()
        // The UUID makes identifiers unique per routine; this only guards a caller passing
        // one routine twice.
        val claimed = HashSet<String>()

        for (routine in routines) {
            if (!routine.enabled) continue
            // Suppress from the FRONT of the day: having trained once, the morning slot is
            // satisfied and the evening one still owed. Dropping the last slot would
            // silence the reminder you need.
            // Front of the TRAINING day (from `DayStamp.ROLLOVER_HOUR`): a 01:00 reminder
            // is the last slot of the evening before, and counting it first would spend the
            // morning's suppression on it.
            val sorted = routine.reminders.toSet().sorted()
            val owed = maxOf(0, sorted.size - maxOf(0, routine.outstandingToday))
            val suppressed = sorted.sortedBy { trainingDayMinute(it) }.take(owed).toSet()
            for (slot in sorted) {
                val id = identifier(routine.id, slot)
                if (!claimed.add(id)) continue
                planned.add(
                    PlannedReminder(
                        identifier = id,
                        title = routine.name,
                        body = L10n.tr("Time for a session."),
                        hour = slot.hour,
                        minute = slot.minute,
                        suppressToday = slot in suppressed,
                    )
                )
            }
        }
        return planned
    }

    /// Minutes since the training day began — the order slots are OWED in.
    private fun trainingDayMinute(slot: ReminderTime): Int =
        Math.floorMod(slot.minutesFromMidnight - DayStamp.ROLLOVER_HOUR * 60, 24 * 60)

    /// Compute the plan and hand it to the scheduler. **Not serialized here**: the one
    /// caller, `TemplateStore` (the door for every replan, boot receiver included), runs
    /// these through its `replanLane`. A lock here would have to be process-global, and one
    /// can be left held by a coroutine whose dispatcher stopped.
    ///
    /// TRANSLATION NOTE: iOS cancels its predecessor Task. The lane cancels nothing and
    /// applies one plan at a time — safer here, because an `AlarmManager` write abandoned
    /// halfway leaves a real alarm behind.
    suspend fun replan(routines: List<RoutinePlanInput>, scheduler: AlarmScheduler) {
        scheduler.apply(requests(routines))
    }
}

/// What talks to the OS. An interface so `ReminderPlannerTests` can assert the plan against
/// a fake.
interface AlarmScheduler {
    /// Add-before-remove: install the new plan FIRST, then drop the remainder. Wiping first
    /// leaves a window with zero reminders, and process death inside it makes that
    /// permanent.
    suspend fun apply(planned: List<ReminderPlanner.PlannedReminder>)
}

/// Records what it was asked to schedule and nothing else — the tests' scheduler, and the
/// app's before any permission exists.
class RecordingAlarmScheduler : AlarmScheduler {
    var applied: List<ReminderPlanner.PlannedReminder> = emptyList()
        private set

    override suspend fun apply(planned: List<ReminderPlanner.PlannedReminder>) {
        applied = planned
    }
}

/// `AlarmManager.setWindow`, a notification channel, and the identifier bookkeeping Android
/// forces because alarms cannot be enumerated.
class AndroidAlarmScheduler(
    context: Context,
    private val settings: RoutineSettings,
    private val zone: ZoneId = ZoneId.systemDefault(),
) : AlarmScheduler {

    private val app = context.applicationContext
    private val alarms = app.getSystemService(AlarmManager::class.java)

    override suspend fun apply(planned: List<ReminderPlanner.PlannedReminder>) {
        // Note what we believe is scheduled but DON'T cancel yet — see
        // `AlarmScheduler.apply`.
        val ours = settings.scheduledReminderIdentifiers

        if (planned.isEmpty()) {
            // All reminders off (or no routines): the plan is genuinely empty, the one path
            // that may clear without checking permission.
            ours.forEach { cancel(it) }
            settings.setScheduledReminderIdentifiers(emptySet())
            return
        }

        // An alarm for a notification we may not post wakes the phone to do nothing. Until
        // `POST_NOTIFICATIONS` is granted (asked on the first Save wanting reminders — see
        // `NotificationPermissionGate`) the plan is not installed, and nothing is
        // cancelled: the user has not said no.
        if (!canPost()) return

        val scheduled = HashSet<String>()
        for (item in planned) {
            schedule(item)
            scheduled.add(item.identifier)
        }

        // Only now drop what the new plan no longer covers.
        (ours - scheduled).forEach { cancel(it) }
        settings.setScheduledReminderIdentifiers(scheduled)
    }

    private fun canPost(): Boolean =
        ContextCompat.checkSelfPermission(app, Manifest.permission.POST_NOTIFICATIONS) ==
            PackageManager.PERMISSION_GRANTED

    private fun schedule(item: ReminderPlanner.PlannedReminder) {
        val at = ReminderAlarms.nextOccurrence(item.hour, item.minute, zone,
            suppressToday = item.suppressToday)
        alarms.setWindow(
            AlarmManager.RTC_WAKEUP,
            at,
            ReminderAlarms.windowMillis,
            ReminderAlarms.pendingIntent(app, item),
        )
    }

    private fun cancel(identifier: String) {
        ReminderAlarms.cancelIntent(app, identifier)?.let {
            alarms.cancel(it)
            it.cancel()
        }
    }
}

/// Shared alarm plumbing — intent shape, window, "when is the next one" — so scheduler and
/// rescheduling receiver cannot disagree.
object ReminderAlarms {
    const val channelId = "reminders"
    const val action = "run.nuri.getagrip.REMINDER"
    const val extraIdentifier = "identifier"
    const val extraTitle = "title"
    const val extraBody = "body"
    const val extraHour = "hour"
    const val extraMinute = "minute"

    /// Fifteen minutes: a habit nudge does not need the exact minute (see the planner's
    /// note on `SCHEDULE_EXACT_ALARM`).
    const val windowMillis: Long = 15 * 60 * 1000L

    /// The first firing of `hour:minute` after `from`, skipped once more when suppressed
    /// AND still inside the current training day.
    ///
    /// **Suppression is about the TRAINING day, not the calendar.** Pushing the slot past
    /// its next calendar occurrence unconditionally is right at 18:00 and wrong after
    /// midnight: a session at 00:30 settles the day that began at 04:00 yesterday, so
    /// tomorrow's 08:00 is the next training day's first reminder and was being skipped
    /// (the same bug existed on iOS). A slot is skipped only when its next firing lands
    /// before the coming rollover.
    fun nextOccurrence(hour: Int, minute: Int, zone: ZoneId, from: LocalDateTime? = null,
                       suppressToday: Boolean = false): Long {
        val now = from ?: LocalDateTime.now(zone)
        var next = LocalDateTime.of(now.toLocalDate(), java.time.LocalTime.of(hour, minute))
        if (!next.isAfter(now)) next = next.plusDays(1)
        if (suppressToday && next.isBefore(trainingDayEnd(now))) next = next.plusDays(1)
        return next.atZone(zone).toInstant().toEpochMilli()
    }

    /// The coming `DayStamp.ROLLOVER_HOUR` in local wall time, where `now`'s training day
    /// ends. Local-time arithmetic, so a DST night moves the hour with the clock.
    fun trainingDayEnd(now: LocalDateTime): LocalDateTime {
        val rollover = now.toLocalDate().atTime(DayStamp.ROLLOVER_HOUR, 0)
        return if (now.isBefore(rollover)) rollover else rollover.plusDays(1)
    }

    fun tomorrow(hour: Int, minute: Int, zone: ZoneId, today: LocalDate): Long =
        LocalDateTime.of(today.plusDays(1), java.time.LocalTime.of(hour, minute))
            .atZone(zone).toInstant().toEpochMilli()

    fun pendingIntent(
        context: Context,
        item: ReminderPlanner.PlannedReminder,
    ): PendingIntent {
        val intent = Intent(context, ReminderReceiver::class.java).apply {
            action = ReminderAlarms.action
            putExtra(extraIdentifier, item.identifier)
            putExtra(extraTitle, item.title)
            putExtra(extraBody, item.body)
            putExtra(extraHour, item.hour)
            putExtra(extraMinute, item.minute)
        }
        return PendingIntent.getBroadcast(
            context,
            item.requestCode,
            intent,
            // UPDATE_CURRENT makes a replan idempotent: re-adding an identifier replaces
            // the alarm in place.
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
    }

    /// NO_CREATE: with nothing scheduled there is nothing to cancel, and minting a
    /// `PendingIntent` to cancel it leaves a fresh one behind on some OEM builds.
    fun cancelIntent(context: Context, identifier: String): PendingIntent? {
        val intent = Intent(context, ReminderReceiver::class.java).apply {
            action = ReminderAlarms.action
        }
        return PendingIntent.getBroadcast(
            context,
            identifier.hashCode(),
            intent,
            PendingIntent.FLAG_NO_CREATE or PendingIntent.FLAG_IMMUTABLE,
        )
    }

    /// **DEFAULT importance, never HIGH.** A habit nudge claiming heads-up banners is how
    /// people turn a whole app's notifications off. The iOS twin is
    /// `interruptionLevel = .active`.
    fun ensureChannel(context: Context) {
        val manager = context.getSystemService(NotificationManager::class.java)
        if (manager.getNotificationChannel(channelId) != null) return
        manager.createNotificationChannel(
            NotificationChannel(
                channelId,
                L10n.tr("Session reminders"),
                NotificationManager.IMPORTANCE_DEFAULT,
            )
        )
    }
}

/// Posts one reminder and schedules the same slot for tomorrow — the replacement for iOS's
/// repeating trigger. Here rather than at launch because a phone unopened for a week must
/// still be reminded; the next replan overwrites this alarm in place, so the two cannot
/// drift.
class ReminderReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action != ReminderAlarms.action) return
        val identifier = intent.getStringExtra(ReminderAlarms.extraIdentifier) ?: return
        val title = intent.getStringExtra(ReminderAlarms.extraTitle) ?: return
        val body = intent.getStringExtra(ReminderAlarms.extraBody)
            ?: L10n.tr("Time for a session.")
        val hour = intent.getIntExtra(ReminderAlarms.extraHour, -1)
        val minute = intent.getIntExtra(ReminderAlarms.extraMinute, -1)
        if (hour < 0 || minute < 0) return

        post(context, identifier, title, body)
        rescheduleTomorrow(context, identifier, title, body, hour, minute)
    }

    private fun post(context: Context, identifier: String, title: String, body: String) {
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) !=
            PackageManager.PERMISSION_GRANTED
        ) return

        ReminderAlarms.ensureChannel(context)
        val open = PendingIntent.getActivity(
            context,
            0,
            Intent(context, MainActivity::class.java)
                .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP),
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
        val notification = NotificationCompat.Builder(context, ReminderAlarms.channelId)
            .setSmallIcon(R.drawable.ic_launcher_monochrome)
            .setContentTitle(title)
            .setContentText(body)
            .setContentIntent(open)
            .setAutoCancel(true)
            .setCategory(Notification.CATEGORY_REMINDER)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .build()
        context.getSystemService(NotificationManager::class.java)
            .notify(identifier.hashCode(), notification)
    }

    private fun rescheduleTomorrow(
        context: Context,
        identifier: String,
        title: String,
        body: String,
        hour: Int,
        minute: Int,
    ) {
        val zone = ZoneId.systemDefault()
        val item = ReminderPlanner.PlannedReminder(identifier, title, body, hour, minute)
        context.getSystemService(AlarmManager::class.java).setWindow(
            AlarmManager.RTC_WAKEUP,
            ReminderAlarms.tomorrow(hour, minute, zone, LocalDate.now(zone)),
            ReminderAlarms.windowMillis,
            ReminderAlarms.pendingIntent(context, item),
        )
    }
}

/// **Asking for `POST_NOTIFICATIONS` belongs to the first Save of a routine that wants
/// reminders, never to launch** — by then the reason is on the previous screen. The twin of
/// `DeviceStore`'s `PermissionGate`: the Activity lends one to the store.
///
/// Nothing installs a gate yet; `TemplateStore.askNotificationPermissionOnce` does its
/// bookkeeping with nobody to ask until the builder wires the dialog.
fun interface NotificationPermissionGate {
    fun request(onResult: (Boolean) -> Unit)
}
